File: elohim/elohim-render/src/shim/node_buffer.js
// elohim-render Node `Buffer` shim: a REAL Uint8Array subclass installed as the
// global `Buffer`.
//
// Why a global (not just a module): the Angular server bundle's bundled `ws`
// evaluates `EMPTY_BUFFER = Buffer.alloc(0)` and `Buffer.from([0,0,255,255])` as
// TOP-LEVEL consts -- at module-eval time, during the SSR render's module-graph
// load. A bare deno_core runtime has no `Buffer` global, so this was the render
// wall (`ReferenceError: Buffer is not defined`). Extension `js` files run at
// isolate init (before any bundle module body), so installing the global here is
// what clears that wall. The `buffer`/`node:buffer` MODULE (served by
// node_builtins) re-exports THIS global, so `import { Buffer }` and the global
// are the same constructor and `x instanceof Buffer` is coherent.
//
// Correctness: `Buffer extends Uint8Array`, so `instanceof Uint8Array` holds and
// `.length`/`.byteOffset`/`.buffer`/`.set` are inherited-correct. utf8 rides
// deno_core's op_encode/op_decode; latin1 rides op_encode_binary_string; base64
// and hex ride the byte-exact Rust ops in node_buffer.rs. See that file's module
// doc for the full used-surface table and the loud-stub rationale.
//
// WARNING: pure ASCII (served through the `js = [dir ...]` macro).
((globalThis) => {
  const core = Deno.core;

  const b64enc = (bytes, urlSafe) =>
    core.ops.op_elohim_base64_encode(bytes, !!urlSafe);
  const b64dec = (s) => core.ops.op_elohim_base64_decode(s);
  const hexenc = (bytes) => core.ops.op_elohim_hex_encode(bytes);
  const hexdec = (s) => core.ops.op_elohim_hex_decode(s);

  // Normalize an encoding label to a canonical supported key, or null if this
  // shim does not implement it (ucs2/utf16le are intentionally absent -- not in
  // the bundle's used surface). Callers throw a LOUD named error on null rather
  // than guessing bytes.
  function normalizeEncoding(enc) {
    if (enc === undefined || enc === null) return "utf8";
    switch (String(enc).toLowerCase()) {
      case "utf8":
      case "utf-8":
        return "utf8";
      case "hex":
        return "hex";
      case "base64":
        return "base64";
      case "base64url":
        return "base64url";
      case "latin1":
      case "binary":
        return "latin1";
      case "ascii":
        return "ascii";
      default:
        return null;
    }
  }

  function unsupportedEncoding(enc) {
    return new TypeError(
      "elohim-render buffer shim: encoding " +
        JSON.stringify(String(enc)) +
        " is not implemented (SSR runtime shim, " +
        "elohim/elohim-render/src/shim/node_buffer.rs). Supported: " +
        "utf8, hex, base64, base64url, latin1/binary, ascii."
    );
  }

  // string -> a plain Uint8Array of the encoded bytes.
  function stringToBytes(str, enc) {
    const key = normalizeEncoding(enc);
    if (key === null) throw unsupportedEncoding(enc);
    const s = String(str);
    switch (key) {
      case "utf8":
        return core.encode(s);
      case "hex":
        return hexdec(s);
      case "base64":
      case "base64url":
        return b64dec(s);
      case "latin1": {
        const out = new Uint8Array(s.length);
        for (let i = 0; i < s.length; i++) out[i] = s.charCodeAt(i) & 0xff;
        return out;
      }
      case "ascii": {
        const out = new Uint8Array(s.length);
        for (let i = 0; i < s.length; i++) out[i] = s.charCodeAt(i) & 0x7f;
        return out;
      }
      default:
        throw unsupportedEncoding(enc);
    }
  }

  // Build a JS string from raw bytes, masking each byte with `mask` (0xff for
  // latin1, 0x7f for ascii). Chunked so a very large buffer does not blow the
  // String.fromCharCode argument-count limit.
  function bytesToCharString(view, mask) {
    let out = "";
    const CHUNK = 8192;
    for (let i = 0; i < view.length; i += CHUNK) {
      const end = Math.min(i + CHUNK, view.length);
      const codes = new Array(end - i);
      for (let j = i; j < end; j++) codes[j - i] = view[j] & mask;
      out += String.fromCharCode.apply(null, codes);
    }
    return out;
  }

  // bytes (a Uint8Array view) -> a string in `enc`.
  function bytesToString(view, enc) {
    const key = normalizeEncoding(enc);
    if (key === null) throw unsupportedEncoding(enc);
    switch (key) {
      case "utf8":
        return core.decode(view);
      case "hex":
        return hexenc(view);
      case "base64":
        return b64enc(view, false);
      case "base64url":
        return b64enc(view, true);
      case "latin1":
        return core.encodeBinaryString(view);
      case "ascii":
        return bytesToCharString(view, 0x7f);
      default:
        throw unsupportedEncoding(enc);
    }
  }

  function isSharedArrayBuffer(v) {
    return (
      typeof SharedArrayBuffer !== "undefined" && v instanceof SharedArrayBuffer
    );
  }

  class Buffer extends Uint8Array {
    // Delegates to Uint8Array: `new Buffer(n)` allocates n zero-filled bytes;
    // `new Buffer(arrayBuffer, off, len)` shares memory (used by species-driven
    // subarray/slice); `new Buffer(uint8arrayOrArray)` copies + octet-truncates.
    // The bundle never calls `new Buffer(...)` directly (Buffer.from/alloc only);
    // this is the internal + @@species constructor.
    constructor(...args) {
      super(...args);
    }

    // --- static: construction ---------------------------------------------

    static from(value, encodingOrOffset, length) {
      if (value === null || value === undefined) {
        throw new TypeError(
          "elohim-render buffer shim: Buffer.from(null/undefined) is invalid"
        );
      }
      if (typeof value === "string") {
        return new Buffer(stringToBytes(value, encodingOrOffset));
      }
      if (value instanceof ArrayBuffer || isSharedArrayBuffer(value)) {
        // Node shares memory with the ArrayBuffer (zero-copy).
        const off = encodingOrOffset === undefined ? 0 : encodingOrOffset >>> 0;
        const len =
          length === undefined ? value.byteLength - off : length >>> 0;
        return new Buffer(value, off, len);
      }
      if (ArrayBuffer.isView(value)) {
        // Copy the BYTES (Node copies from a Buffer/TypedArray). A non-Uint8Array
        // view (Uint16Array, DataView, ...) is byte-reinterpreted, not element-
        // copied, so build a byte view first.
        const bytes =
          value instanceof Uint8Array
            ? value
            : new Uint8Array(value.buffer, value.byteOffset, value.byteLength);
        return new Buffer(bytes);
      }
      if (Array.isArray(value) || typeof value.length === "number") {
        // Octet array-like: super octet-truncates each element (matches Node's
        // `list[i] & 255`).
        return new Buffer(value);
      }
      throw new TypeError(
        "elohim-render buffer shim: Buffer.from(...) unsupported argument type"
      );
    }

    static of(...items) {
      return Buffer.from(items);
    }

    static alloc(size, fill, encoding) {
      const b = new Buffer(size >>> 0); // zero-filled by Uint8Array
      if (fill !== undefined && !(typeof fill === "number" && fill === 0)) {
        b.fill(fill, 0, b.length, encoding);
      }
      return b;
    }

    // Node's allocUnsafe returns UNinitialized memory; a Uint8Array is always
    // zero-filled. Zero-filling is a safe superset (never wrong, only slower),
    // so we intentionally return zeroed memory -- documented divergence.
    static allocUnsafe(size) {
      return new Buffer(size >>> 0);
    }

    static allocUnsafeSlow(size) {
      return new Buffer(size >>> 0);
    }

    static concat(list, totalLength) {
      if (!Array.isArray(list) && typeof list[Symbol.iterator] !== "function") {
        throw new TypeError(
          "elohim-render buffer shim: Buffer.concat expects a list of buffers"
        );
      }
      const items = Array.from(list);
      if (totalLength === undefined) {
        totalLength = 0;
        for (let i = 0; i < items.length; i++) totalLength += items[i].length;
      }
      totalLength = totalLength >>> 0;
      const out = new Buffer(totalLength);
      let pos = 0;
      for (let i = 0; i < items.length && pos < totalLength; i++) {
        let src = items[i];
        const room = totalLength - pos;
        if (src.length > room) src = src.subarray(0, room);
        out.set(src, pos);
        pos += src.length;
      }
      return out;
    }

    static isBuffer(x) {
      return x instanceof Buffer;
    }

    static isEncoding(enc) {
      return typeof enc === "string" && normalizeEncoding(enc) !== null;
    }

    static byteLength(input, encoding) {
      if (typeof input !== "string") {
        if (ArrayBuffer.isView(input)) return input.byteLength;
        if (input instanceof ArrayBuffer || isSharedArrayBuffer(input)) {
          return input.byteLength;
        }
        input = String(input);
      }
      return stringToBytes(input, encoding).length;
    }

    // --- instance ----------------------------------------------------------

    toString(encoding, start, end) {
      const len = this.length;
      let s = start === undefined ? 0 : start | 0;
      let e = end === undefined ? len : end | 0;
      if (s < 0) s = 0;
      if (e > len) e = len;
      if (e < s) e = s;
      const view = s === 0 && e === len ? this : this.subarray(s, e);
      return bytesToString(view, encoding);
    }

    // Node's Buffer.slice SHARES memory (an alias of subarray) -- unlike
    // Uint8Array.prototype.slice, which COPIES. Override to the sharing form so a
    // mutation-through-slice matches Node.
    slice(start, end) {
      return this.subarray(start, end);
    }

    equals(other) {
      if (this === other) return true;
      if (!(other instanceof Uint8Array)) return false;
      if (this.length !== other.length) return false;
      for (let i = 0; i < this.length; i++) {
        if (this[i] !== other[i]) return false;
      }
      return true;
    }

    copy(target, targetStart, sourceStart, sourceEnd) {
      targetStart = targetStart === undefined ? 0 : targetStart | 0;
      sourceStart = sourceStart === undefined ? 0 : sourceStart | 0;
      sourceEnd = sourceEnd === undefined ? this.length : sourceEnd | 0;
      if (sourceEnd < sourceStart) sourceEnd = sourceStart;
      let src = this.subarray(sourceStart, sourceEnd);
      const room = target.length - targetStart;
      if (src.length > room) src = src.subarray(0, Math.max(0, room));
      target.set(src, targetStart);
      return src.length;
    }

    fill(value, start, end, encoding) {
      // Node signatures: fill(value[, offset[, end]][, encoding]).
      if (typeof start === "string") {
        encoding = start;
        start = 0;
        end = this.length;
      } else if (typeof end === "string") {
        encoding = end;
        end = this.length;
      }
      start = start === undefined ? 0 : start | 0;
      end = end === undefined ? this.length : end | 0;
      if (start < 0) start = 0;
      if (end > this.length) end = this.length;
      if (end <= start) return this;

      if (typeof value === "number") {
        Uint8Array.prototype.fill.call(this, value & 0xff, start, end);
        return this;
      }
      let bytes;
      if (typeof value === "string") {
        bytes = stringToBytes(value, encoding);
      } else if (value instanceof Uint8Array) {
        bytes = value;
      } else {
        Uint8Array.prototype.fill.call(this, 0, start, end);
        return this;
      }
      if (bytes.length === 0) {
        Uint8Array.prototype.fill.call(this, 0, start, end);
        return this;
      }
      for (let i = start, j = 0; i < end; i++) {
        this[i] = bytes[j];
        j = j + 1 === bytes.length ? 0 : j + 1;
      }
      return this;
    }

    write(string, offset, length, encoding) {
      // Node signatures: write(string[, offset[, length]][, encoding]).
      if (offset === undefined) {
        offset = 0;
        length = this.length;
      } else if (typeof offset === "string") {
        encoding = offset;
        offset = 0;
        length = this.length;
      } else {
        offset = offset | 0;
        if (typeof length === "string") {
          encoding = length;
          length = this.length - offset;
        } else if (length === undefined) {
          length = this.length - offset;
        } else {
          length = length | 0;
        }
      }
      const bytes = stringToBytes(String(string), encoding);
      const n = Math.min(length, bytes.length, this.length - offset);
      if (n <= 0) return 0;
      this.set(bytes.subarray(0, n), offset);
      return n;
    }

    toJSON() {
      return { type: "Buffer", data: Array.prototype.slice.call(this) };
    }
  }

  // Off-boot-path binary wire-protocol read/write family: LOUD, NAMED stubs on
  // Buffer.prototype (see node_buffer.rs module doc). Used by `ws` frame framing
  // and protobuf varint at CONNECTION time, never during a single-shot SSR
  // render. A guessed offset/return impl would be silent-wrong on a live wire, so
  // the honest bound is a throw naming the method + this file. Installed on the
  // prototype so a call on a real Buffer instance is loud; a call on a non-Buffer
  // (a custom protobuf Writer) uses that object's own method and never reaches
  // here.
  const LOUD_MEMBERS = [
    "writeUInt8", "writeUInt16LE", "writeUInt16BE", "writeUInt32LE",
    "writeUInt32BE", "writeUIntLE", "writeUIntBE",
    "writeInt8", "writeInt16LE", "writeInt16BE", "writeInt32LE",
    "writeInt32BE", "writeIntLE", "writeIntBE",
    "writeBigUInt64LE", "writeBigUInt64BE", "writeBigInt64LE", "writeBigInt64BE",
    "writeFloatLE", "writeFloatBE", "writeDoubleLE", "writeDoubleBE",
    "readUInt8", "readUInt16LE", "readUInt16BE", "readUInt32LE",
    "readUInt32BE", "readUIntLE", "readUIntBE",
    "readInt8", "readInt16LE", "readInt16BE", "readInt32LE",
    "readInt32BE", "readIntLE", "readIntBE",
    "readBigUInt64LE", "readBigUInt64BE", "readBigInt64LE", "readBigInt64BE",
    "readFloatLE", "readFloatBE", "readDoubleLE", "readDoubleBE",
    "swap16", "swap32", "swap64",
  ];
  for (let i = 0; i < LOUD_MEMBERS.length; i++) {
    const name = LOUD_MEMBERS[i];
    if (typeof Buffer.prototype[name] === "function") continue;
    Object.defineProperty(Buffer.prototype, name, {
      value: function loudBufferMember() {
        throw new Error(
          "elohim-render buffer shim: " +
            name +
            "() is not implemented (SSR runtime shim, " +
            "elohim/elohim-render/src/shim/node_buffer.rs). It is an off-boot-path " +
            "binary wire-protocol method (ws framing / protobuf varint, called at " +
            "connection time, not during SSR render). Implement it deliberately " +
            "(byte-exact via DataView) if a render path proves it needs it."
        );
      },
      writable: true,
      configurable: true,
      enumerable: false,
    });
  }

  if (typeof globalThis.Buffer === "undefined") {
    globalThis.Buffer = Buffer;
  }
})(globalThis);

File: elohim/elohim-render/src/shim/node_buffer.rs
//! Rust half of the Node `Buffer` shim.
//!
//! The Angular server bundle (via bundled `ws` / `multiformats` / libp2p code)
//! uses a Node `Buffer` in two shapes, BOTH of which must resolve to ONE real
//! class:
//!
//! - a **global** `Buffer` -- e.g. `ws`'s `EMPTY_BUFFER = Buffer.alloc(0)` and
//!   `Buffer.from([0,0,255,255])` evaluate at module-eval time (top-level consts)
//!   during the SSR render's module-graph load. A bare deno_core runtime has no
//!   `Buffer` global, so this was the render wall
//!   (`ReferenceError: Buffer is not defined` at the ws chunk).
//! - a named export `Buffer` from the `buffer` / `node:buffer` module -- e.g.
//!   `import { Buffer } from "buffer"` in multiformats/libp2p chunks. That import
//!   is served by [`crate::shim::node_builtins`], whose `("buffer","Buffer")`
//!   member is now REAL and simply re-exports `globalThis.Buffer` (one source of
//!   truth: the global installed by the JS half here).
//!
//! # Architecture: a sibling ext (like [`crate::shim::node_crypto`])
//!
//! The GLOBAL must exist at runtime-init time (before ANY bundle module body runs
//! -- the ws const is module-eval-time). deno_core extension `js` files run at
//! isolate construction, so the global is installed there: [`node_buffer.js`] is
//! served via `js = [dir "src/shim", "node_buffer.js"]` (NOT as a loader module),
//! exactly the init-time install path [`crate::shim::node_globals`] uses. The
//! `buffer` MODULE re-exports that global rather than defining a second class, so
//! `import { Buffer }` and the global `Buffer` are the SAME constructor and
//! `x instanceof Buffer` is coherent across both surfaces.
//!
//! # Why Rust ops back base64 / hex (but not utf8)
//!
//! `Buffer` is a `Uint8Array` subclass; utf8 round-trips through deno_core's
//! existing `Deno.core.encode` / `Deno.core.decode` (op_encode/op_decode) and
//! latin1 through `Deno.core.encodeBinaryString`. deno_core core provides NO
//! base64 or hex primitive (verified against deno_core 0.339 `01_core.js`: only
//! encode/decode/encodeBinaryString are exposed; no `atob`/`btoa`). Rather than
//! hand-roll a base64 alphabet in JS (subtle padding/whitespace bugs are exactly
//! the silent-wrong class this rung is meant to avoid), the four codecs are
//! byte-exact pure-Rust ops, unit-tested here independent of V8 -- mirroring how
//! `node_crypto`'s `createHash` is a real sha2 op, not a JS fabrication.
//!
//! # Used-surface discipline (mirrors node_builtins / node_crypto)
//!
//! Implemented for REAL is exactly the surface the bundle exercises at
//! module-eval / SSR-render time:
//!
//! Static: `from(string,enc | ArrayBuffer[,off,len] | TypedArray/Buffer | Array)`,
//! `alloc(size[,fill[,enc]])`, `allocUnsafe`, `allocUnsafeSlow`, `concat`,
//! `isBuffer`, `byteLength`, `isEncoding`.
//! Instance: `toString(enc[,start[,end]])`, `slice`/`subarray` (memory-SHARING,
//! per Node -- NOT Uint8Array's copying `slice`), `equals`, `copy`, `fill`,
//! `write`, `toJSON`, and the inherited `length` / `byteOffset` / `buffer`.
//! Encodings: `utf8`/`utf-8`, `hex`, `base64`, `base64url`, `latin1`/`binary`,
//! `ascii`. (`ucs2`/`utf16le` are NOT in the bundle's used surface -- absent, per
//! the honesty rule.)
//!
//! LOUD, NAMED stubs (used by the bundle but off the SSR boot path -- the binary
//! wire-protocol read/write integer family lives in `ws` frame framing
//! (`Sender.frame` -> `writeUInt16BE`) and protobuf varint code, which runs at
//! CONNECTION time, never during a single-shot SSR render -- the same class as
//! `zlib`'s connection-time loud stubs). Getting Node's exact `writeUIntBE`
//! offset/return semantics subtly wrong would be silent-wrong on a live wire, so
//! the honest bound is a throw naming the method + this file, NOT a guessed impl.
//! Installed on `Buffer.prototype` so a call on a real `Buffer` instance is loud;
//! a call on a non-Buffer (a custom protobuf `Writer`) uses that object's own
//! method and never consults this prototype.
//!
//! # WARNING: pure ASCII
//!
//! `node_buffer.js` is served through the `js = [dir ...]` macro, which panics at
//! const-eval on any non-ASCII byte. Keep every literal ASCII.

use deno_core::op2;

/// The standard and URL-safe base64 alphabets (RFC 4648).
const B64_STD: &[u8; 64] = b"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
const B64_URL: &[u8; 64] = b"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_";
const HEX_LC: &[u8; 16] = b"0123456789abcdef";

/// Encode `data` as base64. `url_safe` selects the `-_` alphabet and, per Node's
/// `base64url`, emits NO `=` padding; standard base64 pads to a multiple of 4.
///
/// Pure (V8-free) so it is unit-tested directly below.
fn base64_encode(data: &[u8], url_safe: bool) -> String {
    let alphabet = if url_safe { B64_URL } else { B64_STD };
    let mut out = String::with_capacity(data.len().div_ceil(3) * 4);
    for chunk in data.chunks(3) {
        let b0 = chunk[0] as u32;
        let b1 = if chunk.len() > 1 { chunk[1] as u32 } else { 0 };
        let b2 = if chunk.len() > 2 { chunk[2] as u32 } else { 0 };
        let n = (b0 << 16) | (b1 << 8) | b2;
        out.push(alphabet[((n >> 18) & 63) as usize] as char);
        out.push(alphabet[((n >> 12) & 63) as usize] as char);
        match chunk.len() {
            1 => {
                if !url_safe {
                    out.push('=');
                    out.push('=');
                }
            }
            2 => {
                out.push(alphabet[((n >> 6) & 63) as usize] as char);
                if !url_safe {
                    out.push('=');
                }
            }
            _ => {
                out.push(alphabet[((n >> 6) & 63) as usize] as char);
                out.push(alphabet[(n & 63) as usize] as char);
            }
        }
    }
    out
}

/// Decode a base64 string. Lenient like Node: accepts BOTH the standard (`+/`)
/// and URL-safe (`-_`) alphabets (so one decoder serves `base64` and
/// `base64url`), treats `=` as an end-of-data terminator, and skips any other
/// non-alphabet byte (whitespace/newlines in wrapped base64). Well-formed inputs
/// -- the content hashes / keys the bundle decodes -- round-trip exactly.
fn base64_decode(s: &str) -> Vec<u8> {
    fn val(c: u8) -> Option<u8> {
        match c {
            b'A'..=b'Z' => Some(c - b'A'),
            b'a'..=b'z' => Some(c - b'a' + 26),
            b'0'..=b'9' => Some(c - b'0' + 52),
            b'+' | b'-' => Some(62),
            b'/' | b'_' => Some(63),
            _ => None,
        }
    }
    let mut out = Vec::with_capacity(s.len() / 4 * 3 + 3);
    let mut acc: u32 = 0;
    let mut bits: u8 = 0;
    for &c in s.as_bytes() {
        if c == b'=' {
            break;
        }
        let v = match val(c) {
            Some(v) => v,
            None => continue,
        };
        acc = (acc << 6) | v as u32;
        bits += 6;
        if bits >= 8 {
            bits -= 8;
            out.push((acc >> bits) as u8);
        }
    }
    out
}

/// Encode `data` as lowercase hex (Node's `hex` encoding output).
fn hex_encode(data: &[u8]) -> String {
    let mut out = String::with_capacity(data.len() * 2);
    for &b in data {
        out.push(HEX_LC[(b >> 4) as usize] as char);
        out.push(HEX_LC[(b & 0x0f) as usize] as char);
    }
    out
}

/// Decode a hex string. Like Node: decode whole byte-pairs, TRUNCATE at the first
/// non-hex character or a trailing odd nibble (never throw -- Node's `hex`
/// decoder silently truncates).
fn hex_decode(s: &str) -> Vec<u8> {
    fn nib(c: u8) -> Option<u8> {
        match c {
            b'0'..=b'9' => Some(c - b'0'),
            b'a'..=b'f' => Some(c - b'a' + 10),
            b'A'..=b'F' => Some(c - b'A' + 10),
            _ => None,
        }
    }
    let bytes = s.as_bytes();
    let mut out = Vec::with_capacity(bytes.len() / 2);
    let mut i = 0;
    while i + 1 < bytes.len() {
        match (nib(bytes[i]), nib(bytes[i + 1])) {
            (Some(h), Some(l)) => {
                out.push((h << 4) | l);
                i += 2;
            }
            _ => break,
        }
    }
    out
}

/// Base64-encode `data`; `url_safe` picks the URL-safe alphabet with no padding.
/// Backs the JS shim's `toString("base64" | "base64url")`.
#[op2]
#[string]
fn op_elohim_base64_encode(#[buffer] data: &[u8], url_safe: bool) -> String {
    base64_encode(data, url_safe)
}

/// Base64-decode `s` (lenient, both alphabets). Backs `Buffer.from(s, "base64" |
/// "base64url")`.
#[op2]
#[buffer]
fn op_elohim_base64_decode(#[string] s: &str) -> Vec<u8> {
    base64_decode(s)
}

/// Hex-encode `data` (lowercase). Backs the JS shim's `toString("hex")`.
#[op2]
#[string]
fn op_elohim_hex_encode(#[buffer] data: &[u8]) -> String {
    hex_encode(data)
}

/// Hex-decode `s` (truncating on the first invalid/odd nibble). Backs
/// `Buffer.from(s, "hex")`.
#[op2]
#[buffer]
fn op_elohim_hex_decode(#[string] s: &str) -> Vec<u8> {
    hex_decode(s)
}

deno_core::extension!(
    node_buffer_ext,
    ops = [
        op_elohim_base64_encode,
        op_elohim_base64_decode,
        op_elohim_hex_encode,
        op_elohim_hex_decode,
    ],
    js = [dir "src/shim", "node_buffer.js"],
);

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn base64_round_trips_all_pad_lengths() {
        // 0,1,2 trailing bytes exercise the three padding cases.
        for input in [&b""[..], b"f", b"fo", b"foo", b"foob", b"fooba", b"foobar"] {
            let enc = base64_encode(input, false);
            assert_eq!(base64_decode(&enc), input, "std round-trip for {input:?}");
        }
    }

    #[test]
    fn base64_known_vectors() {
        // RFC 4648 test vectors.
        assert_eq!(base64_encode(b"", false), "");
        assert_eq!(base64_encode(b"f", false), "Zg==");
        assert_eq!(base64_encode(b"fo", false), "Zm8=");
        assert_eq!(base64_encode(b"foo", false), "Zm9v");
        assert_eq!(base64_encode(b"foob", false), "Zm9vYg==");
        assert_eq!(base64_encode(b"fooba", false), "Zm9vYmE=");
        assert_eq!(base64_encode(b"foobar", false), "Zm9vYmFy");
    }

    #[test]
    fn base64url_no_padding_and_urlsafe_alphabet() {
        // 0xff,0xef,0xfe -> "/+/+" in std, "_-_-"... build a case that yields +//.
        // Bytes chosen so std encoding contains + and /.
        let data = [0xfb_u8, 0xff, 0xbf];
        let std = base64_encode(&data, false);
        let url = base64_encode(&data, true);
        assert!(std.contains('+') || std.contains('/'), "std uses +/: {std}");
        assert!(
            !url.contains('+') && !url.contains('/'),
            "url-safe avoids +/: {url}"
        );
        assert!(!url.contains('='), "base64url has no padding: {url}");
        // Both decode back to the same bytes via the lenient decoder.
        assert_eq!(base64_decode(&std), data);
        assert_eq!(base64_decode(&url), data);
    }

    #[test]
    fn base64_decode_skips_whitespace() {
        assert_eq!(base64_decode("Zm9v\nYmFy"), b"foobar");
        assert_eq!(base64_decode("Zm9v YmFy"), b"foobar");
    }

    #[test]
    fn hex_round_trips_and_is_lowercase() {
        let data = [0x00_u8, 0x0f, 0xa5, 0xff, 0xde, 0xad, 0xbe, 0xef];
        let enc = hex_encode(&data);
        assert_eq!(enc, "000fa5ffdeadbeef");
        assert_eq!(hex_decode(&enc), data);
    }

    #[test]
    fn hex_decode_accepts_uppercase_and_truncates_odd_and_invalid() {
        assert_eq!(hex_decode("DEADBEEF"), [0xde, 0xad, 0xbe, 0xef]);
        // Odd trailing nibble is dropped (Node truncates the incomplete byte).
        assert_eq!(hex_decode("dead0"), [0xde, 0xad]);
        // Truncates at the first non-hex character.
        assert_eq!(hex_decode("deadZZ00"), [0xde, 0xad]);
        assert_eq!(hex_decode("z"), Vec::<u8>::new());
    }
}
